### PR TITLE
-better handling of multiple handles being returned in personalization:decisions responses

### PR DESCRIPTION
**Pull request by sbenedicadb**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: AEPMessaging/Sources/Messaging.swift [hunk 1]
```diff
@@ -160,20 +161,28 @@ public class Messaging: NSObject, Extension {
...
-
-        rulesEngine.setPropositionsCache(propositions)
-        Log.trace(label: MessagingConstants.LOG_TAG, "Loading in-app message definition from network response.")
-        rulesEngine.loadPropositions(propositions)
+        
+        // if this is an event for a new request, purge cache and update lastProcessedRequestEventId
+        var clearExistingRules = false
+        if lastProcessedRequestEventId != event.requestEventId {
+            clearExistingRules = true
+            rulesEngine.clearPropositionsCache()
```
[the remark this answers is not in the document]
Comment (sbenedicadb) [reply AEPMessaging/Sources/Messaging.swift:179]: yep.  the code in MessagingRulesEngine was refactored and it now makes sense to have it handled there.  i'll move it.

File: AEPMessaging/Sources/Messaging.swift [hunk 2]
```diff
@@ -160,20 +161,27 @@ public class Messaging: NSObject, Extension {
...
+        
+        // if this is an event for a new request, purge cache and update lastProcessedRequestEventId
+        var clearExistingRules = false
+        if lastProcessedRequestEventId != event.requestEventId {
+            clearExistingRules = true            
+            lastProcessedRequestEventId = event.requestEventId
+        }
+                 
+        Log.trace(label: MessagingConstants.LOG_TAG, "Loading in-app message definitions from network response.")
+        rulesEngine.addPropositionsToCache(event.payload)
```
Comment (emdobrin) [inline AEPMessaging/Sources/Messaging.swift:183]: I think [event.payload](https://github.com/adobe/aepsdk-messaging-ios/blob/00348ab14c41be3f0b23a231881d58f6b3dd6dc2/AEPMessaging/Sources/Event%2BMessaging.swift#L217) needs to filter just the prepositions relevant for the requested scope(s) before passing them to addPropositionsToCache

File: AEPMessaging/Tests/UnitTests/MessagingTests.swift [hunk 3]
```diff
@@ -161,20 +160,50 @@ class MessagingTests: XCTestCase {
...
+        messaging.setLastProcessedRequestEventId("oldEventId")
+        let event = Event(name: "Test Offer Notification Event", type: EventType.edge,
+                          source: MessagingConstants.Event.Source.PERSONALIZATION_DECISIONS, data: getOfferEventData())
+
+        // test
+        mockRuntime.simulateComingEvents(event)
+
+        // verify
+        XCTAssertTrue(mockMessagingRulesEngine.loadPropositionsCalled)        
+        XCTAssertEqual(true, mockMessagingRulesEngine.paramLoadPropositionsClearExisting)
```
Comment (swarna04) [inline AEPMessaging/Tests/UnitTests/MessagingTests.swift:193]: `mockMessagingRulesEngine.paramLoadPropositionsPersistChanges` should be `true` here as well.

File: AEPMessaging/Tests/TestHelpers/MockMessagingRulesEngine.swift [hunk 4]
```diff
@@ -44,9 +44,13 @@ class MockMessagingRulesEngine: MessagingRulesEngine {
 
     var loadPropositionsCalled = false
     var paramLoadPropositionsPropositions: [PropositionPayload]?
-    override func loadPropositions(_ propositions: [PropositionPayload]) {
+    var paramLoadPropositionsClearExisting: Bool?
+    var paramLoadPropositionsPersistChanges: Bool?
```
Comment (swarna04) [inline AEPMessaging/Tests/TestHelpers/MockMessagingRulesEngine.swift:48]: These params `paramLoadPropositionsClearExisting` and `paramLoadPropositionsPersistChanges ` do not need to be optional.

File: AEPMessaging/Tests/UnitTests/MessagingRulesEngine+CachingTests.swift [hunk 5]
```diff
@@ -162,37 +162,14 @@ class MessagingRulesEngineCachingTests: XCTestCase {
...
-    func testCachePropositionsSetCache() throws {
+    func testCachePropositionsAddCache() throws {
         // setup
         let propString: String = JSONFileLoader.getRulesStringFromFile("showOnceRule")
         let decoder = JSONDecoder()
         let propositions = try decoder.decode([PropositionPayload].self, from: propString.data(using: .utf8)!)
         
         // test
-        messagingRulesEngine.setPropositionsCache(propositions)
+        messagingRulesEngine.addPropositionsToCache(propositions)
```
Comment (swarna04) [inline AEPMessaging/Tests/UnitTests/MessagingRulesEngine+CachingTests.swift:172]: In the tests for `addPropositionsToCache`, we should also check that the `inMemoryPropositions` are updated.

File: AEPMessaging/Tests/UnitTests/MessagingRulesEngineTests.swift [hunk 6]
```diff
@@ -65,7 +65,21 @@ class MessagingRulesEngineTests: XCTestCase {
         let propositions = try decoder.decode([PropositionPayload].self, from: propString.data(using: .utf8)!)
         
         // test
-        messagingRulesEngine.loadPropositions(propositions)
+        messagingRulesEngine.loadPropositions(propositions, clearExisting: false)
```
Comment (swarna04) [inline AEPMessaging/Tests/UnitTests/MessagingRulesEngineTests.swift:68]: In the `loadPropositions` tests here, we should also check updates to

- inMemoryPropositions
- propositionInfo
- whether changes are persisted with a call to addPropositionsToCache (if applicable)

File: AEPMessaging/Sources/MessagingRulesEngine.swift [hunk 7]
```diff
@@ -44,45 +45,53 @@ class MessagingRulesEngine {
...
-            return
+        if clearExisting {
+            inMemoryPropositions.removeAll()
+            cachePropositions(nil)
+            propositionInfo = tempPropInfo
+            rulesEngine.replaceRules(with: rules)
+            Log.debug(label: MessagingConstants.LOG_TAG, "Successfully loaded \(rules.count) message(s) into the rules engine.")
+        } else {
+            propositionInfo.merge(tempPropInfo) { _, new in new }
+            rulesEngine.addRules(rules)
```
Comment (swarna04) [inline AEPMessaging/Sources/MessagingRulesEngine.swift:86]: Is this a no-op if invoked with empty rules ([]) or we need to make a prior check?

**Review by swarna04 — CHANGES_REQUESTED** [review]

File: AEPMessaging/Sources/Messaging.swift [hunk 8]
```diff
@@ -160,20 +161,27 @@ public class Messaging: NSObject, Extension {
...
             // either this isn't the type of response we are waiting for, or it's not a response for our request
             return
         }
-
-        guard let propositions = event.payload, !propositions.isEmpty, event.scope == appSurface else {
-            Log.debug(label: MessagingConstants.LOG_TAG, "Payload for in-app messages was empty. Clearing in-app messages from cache and persistence.")
-            rulesEngine.clearPropositions()                        
+        
+        // quick out if we have a scope (implying payload is not empty) and the scope doesn't match our app surface
+        if event.scope != nil && event.scope != appSurface {
```
[the remark this answers is not in the document]
Comment (swarna04) [reply AEPMessaging/Sources/Messaging.swift:170]: I agree here that if we want to verify the scope then just checking from the first payload (event.scope) doesn't fit the bill.

File: AEPMessaging/Sources/MessagingRulesEngine.swift (hunk 7 above)
Comment (sbenedicadb) [reply AEPMessaging/Sources/MessagingRulesEngine.swift:89]: if we aren't clearing existing and there are no new rules, it is essentially a no-op

File: AEPMessaging/Tests/TestHelpers/MockMessagingRulesEngine.swift (hunk 4 above)
Comment (sbenedicadb) [reply AEPMessaging/Tests/TestHelpers/MockMessagingRulesEngine.swift:48]: i prefer to leave them optional so they will be nil if this mock method is not called

**Review by swarna04 — APPROVED** [review]

**Pull request merged**